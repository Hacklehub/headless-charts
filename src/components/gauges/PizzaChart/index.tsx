/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/no-unused-vars */

import { arc, interpolateNumber, pie, scaleLinear, transition } from 'd3';
import { defaultChartClassNames, mergeTailwindClasses } from '../../../utils';
import { useCallback, useEffect } from 'react';
import useTooltip, { TooltipObjectType } from '../../../hooks/useTooltip';

import { select } from 'd3-selection';

export interface PizzaChartProps {
  data: { [key: string]: any };
  id: string;
  className?: string;
  margin?: {
    top: number;
    bottom: number;
    left: number;
    right: number;
  };
  max?: number;
  min?: number;
  paddingAngle?: number;
  cornerRadius?: number;
  tooltip?: TooltipObjectType;
  drawing?: {
    duration?: number;
    delay?: number;
  };
  metrics: {
    key: string;
    className?: string;
    classNameBackground?: string;
  }[];
}

const PizzaChart = ({
  data = {},
  className,
  id,
  margin = { top: 20, bottom: 20, left: 20, right: 20 },
  max = 1,
  min = 0,
  drawing = undefined,
  paddingAngle = 2,
  cornerRadius = 0,
  metrics = [],
  tooltip,
}: PizzaChartProps) => {
  const { onMouseOver, onMouseMove, onMouseLeave } = useTooltip({
    tooltip,
    defaultHtml: (d: any) =>
      `<div class="text-gray-800">${metrics[d.index].key} = ${
        data[metrics[d.index].key]
      }</div>`,
  });
  const refreshChart = useCallback(() => {
    const svg = select(`#${id}`);

    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0];
    const height = +svg.style('height').split('px')[0];

    const chartArea = [
      width - margin.left - margin.right,
      height - margin.top - margin.bottom,
    ];

    const maxRadius =
      Math.min(
        width - margin.left - margin.right,
        height - margin.top - margin.bottom
      ) / 2;

    const g = svg
      .append('g')
      .attr(
        'transform',
        `translate(${margin.left + chartArea[0] / 2},${
          margin.top + chartArea[1] / 2
        })`
      );

    const radiusScale = scaleLinear().domain([min, max]).range([0, maxRadius]);

    const pieFn = pie();

    // console.log(pieFn(metrics.map((d: any) => 1)));

    const outsideCircles = g.append('g');

    outsideCircles
      .selectAll('path')
      .data(pieFn(metrics.map(() => 1)))
      .enter()
      .append('path')
      .attr('d', (d: any) =>
        arc()
          .innerRadius(0)
          .outerRadius(maxRadius)
          .padAngle((paddingAngle / 360) * (2 * Math.PI))
          .cornerRadius(cornerRadius)(d)
      )
      .attr('class', (d: any) =>
        mergeTailwindClasses(
          'fill-gray-200',
          metrics[d.index].classNameBackground
        )
      )
      .on('mouseenter', onMouseOver)
      .on('mousemove', onMouseMove)
      .on('mouseleave', onMouseLeave);

    const pathG = g.append('g');
    const arcs = pathG
      .selectAll('path')
      .data(pieFn(metrics.map(() => 1)))
      .join('path')
      .attr('d', (d: any) =>
        arc()
          .innerRadius(0)
          .outerRadius((metric: any) =>
            drawing?.duration ? 0 : radiusScale(data[metrics[metric.index].key])
          )
          .padAngle((paddingAngle / 360) * (2 * Math.PI))
          .cornerRadius(cornerRadius)(d)
      )
      .attr('class', (metric: any) =>
        mergeTailwindClasses('fill-gray-800', metrics[metric.index].className)
      )
      .on('mouseenter', onMouseOver)
      .on('mousemove', onMouseMove)
      .on('mouseleave', onMouseLeave);

    transition();

    if (drawing?.duration) {
      arcs
        .transition()
        .duration(drawing.duration)
        .delay((_, i: number) => i * (drawing?.delay || 0))
        // @ts-ignore
        .attrTween('d', (d: any) => {
          const interpolationFn = interpolateNumber(
            0,
            radiusScale(data[metrics[d.index].key])
          );

          return (t: any) =>
            arc()
              .innerRadius(0)
              .outerRadius(() => interpolationFn(t))
              .padAngle((paddingAngle / 360) * (2 * Math.PI))
              .cornerRadius(cornerRadius)(d);
        });
    }
  }, [
    cornerRadius,
    data,
    id,
    margin,
    max,
    metrics,
    min,
    onMouseLeave,
    onMouseMove,
    onMouseOver,
    paddingAngle,
    drawing,
  ]);

  useEffect(() => {
    refreshChart();
  }, [refreshChart]);

  return (
    <svg
      id={id}
      className={mergeTailwindClasses(className, defaultChartClassNames)}
    />
  );
};

export default PizzaChart;
