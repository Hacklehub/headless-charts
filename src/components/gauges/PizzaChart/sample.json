{
  "name": "A",
  "metric1": 0.5,
  "metric2": 0.4,
  "metric3": 0.3,
  "metric4": 0.55,
  "metric5": 0.4,
  "metric6": 0.6,
  "metric7": 0.7
}
