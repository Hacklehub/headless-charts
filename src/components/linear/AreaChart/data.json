[
  {
    "year": 2012,
    "macbook": 23.22,
    "services": 12.89,
    "wearables": 10.76,
    "ipad": 30.94,
    "iphone": 78.69
  },
  {
    "year": 2013,
    "macbook": 21.48,
    "services": 16.05,
    "wearables": 10.12,
    "ipad": 31.98,
    "iphone": 91.28
  },
  {
    "year": 2014,
    "macbook": 24.08,
    "services": 18.06,
    "wearables": 8.379,
    "ipad": 30.28,
    "iphone": 101.99
  },
  {
    "year": 2015,
    "macbook": 25.47,
    "services": 19.91,
    "wearables": 10.07,
    "ipad": 23.23,
    "iphone": 155.04
  },
  {
    "year": 2016,
    "macbook": 22.83,
    "services": 24.35,
    "wearables": 11.13,
    "ipad": 20.63,
    "iphone": 136.7
  },
  {
    "year": 2017,
    "macbook": 25.57,
    "services": 32.7,
    "wearables": 12.83,
    "ipad": 18.8,
    "iphone": 139.34
  },

  {
    "year": 2018,
    "macbook": 25.2,
    "services": 39.75,
    "wearables": 17.38,
    "ipad": 18.38,
    "iphone": 164.89
  },
  {
    "year": 2019,
    "macbook": 25.74,
    "services": 46.29,
    "wearables": 24.48,
    "ipad": 21.28,
    "iphone": 142.38
  },
  {
    "year": 2020,
    "macbook": 28.62,
    "services": 53.77,
    "wearables": 30.62,
    "ipad": 23.72,
    "iphone": 137.78
  },

  {
    "year": 2021,
    "macbook": 35.19,
    "services": 68.42,
    "wearables": 38.37,
    "ipad": 31.86,
    "iphone": 191.97
  },
  {
    "year": 2022,
    "macbook": 40.18,
    "services": 78.13,
    "wearables": 41.24,
    "ipad": 29.29,
    "iphone": 205.49
  }
]
