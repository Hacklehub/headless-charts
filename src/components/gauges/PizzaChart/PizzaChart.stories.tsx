import { Meta, StoryObj } from '@storybook/react';

import PizzaChart from '.';
import data from './sample.json';

export default {
  title: 'Gauge/PizzaChart',
  component: PizzaChart,
} as Meta;

type Story = StoryObj<typeof PizzaChart>;

export const Default: Story = {
  args: {
    id: 'pizza-chart',
    data,
    metrics: [
      {
        key: 'metric1',
      },
      {
        key: 'metric2',
      },
      {
        key: 'metric3',
      },
      {
        key: 'metric4',
      },
      {
        key: 'metric5',
      },
      {
        key: 'metric6',
      },
      {
        key: 'metric7',
      },
    ],
  },
};

export const WithCustomColors: Story = {
  args: {
    ...Default.args,
    id: 'pizza-chart-custom-colors',
    metrics: [
      {
        key: 'metric1',
        className: 'fill-purple-900',
        classNameBackground: 'fill-gray-200',
      },
      {
        key: 'metric2',
        className: 'fill-purple-800',
        classNameBackground: 'fill-gray-200',
      },
      {
        key: 'metric3',
        className: 'fill-purple-700',
        classNameBackground: 'fill-gray-200',
      },
      {
        key: 'metric4',
        className: 'fill-purple-600',
        classNameBackground: 'fill-gray-200',
      },
      {
        key: 'metric5',
        className: 'fill-purple-500',
        classNameBackground: 'fill-gray-200',
      },
      {
        key: 'metric6',
        className: 'fill-pink-500',
        classNameBackground: 'fill-gray-200',
      },
      {
        key: 'metric7',
        className: 'fill-pink-600',
        classNameBackground: 'fill-gray-200',
      },
    ],
  },
};

export const Drawing: Story = {
  args: {
    ...Default.args,
    id: 'pizza-chart-drawing',
    drawing: {
      duration: 1000,
    },
  },
};

export const DrawingWithDelay: Story = {
  args: {
    ...Default.args,
    id: 'pizza-chart-drawing-delay',
    drawing: {
      duration: 1000,
      delay: 50,
    },
  },
};
