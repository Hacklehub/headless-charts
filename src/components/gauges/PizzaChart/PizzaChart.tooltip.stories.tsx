/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Meta, StoryObj } from '@storybook/react';

import PizzaChart from '.';
import data from './sample.json';

const metrics = [
  {
    key: 'metric1',
  },
  {
    key: 'metric2',
  },
  {
    key: 'metric3',
  },
  {
    key: 'metric4',
  },
  {
    key: 'metric5',
  },
  {
    key: 'metric6',
  },
  {
    key: 'metric7',
  },
];

export default {
  title: 'Gauge/PizzaChart/Tooltip',
  component: PizzaChart,
} as Meta;

type Story = StoryObj<typeof PizzaChart>;

export const WithTooltip: Story = {
  args: {
    id: 'pizza-chart-tooltip',
    data,
    tooltip: {},
    metrics,
  },
};

export const WithCustomColors: Story = {
  args: {
    ...WithTooltip.args,
    id: 'pizza-chart-tooltip-custom-colors',
    tooltip: {
      className: 'bg-white border-2 border-purple-700 p-2 rounded',
    },
  },
};

export const WithCustomHtml: Story = {
  args: {
    ...WithTooltip.args,
    id: 'pizza-chart-tooltip-custom-html',
    tooltip: {
      html: (d: { index: number; data: number }) =>
        `<div class="bg-white border-2 border-purple-700 p-2 rounded">
        ${metrics[d.index].key} was scored at ${data[metrics[d.index].key]}.
        </div>`,
    },
  },
};
