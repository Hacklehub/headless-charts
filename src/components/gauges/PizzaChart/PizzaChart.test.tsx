import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import PizzaChart from '.';
import data from './sample.json';

describe('PizzaChart', () => {
  it('renders PizzaChart component for snapshot test', () => {
    render(<PizzaChart id='pizza-chart'
      data={data}
      metrics={[
        {
          key: 'metric1',
        },
        {
          key: 'metric2',
        },
        {
          key: 'metric3',
        },
        {
          key: 'metric4',
        },
        {
          key: 'metric5',
        },
        {
          key: 'metric6',
        },
        {
          key: 'metric7',
        },
      ]} />);
    const pizzaChart = screen.getByTestId('pizza-chart');
    expect(pizzaChart).toMatchSnapshot();
  });

  it('renders PizzaChart component', () => {
    render(<PizzaChart id='pizza-chart'
      data={data}
      metrics={[
        {
          key: 'metric1',
        },
        {
          key: 'metric2',
        },
        {
          key: 'metric3',
        },
        {
          key: 'metric4',
        },
        {
          key: 'metric5',
        },
        {
          key: 'metric6',
        },
        {
          key: 'metric7',
        },
      ]} />);
    const pizzaChart = screen.getByTestId('pizza-chart');
    expect(pizzaChart).toBeInTheDocument();
  });

})
