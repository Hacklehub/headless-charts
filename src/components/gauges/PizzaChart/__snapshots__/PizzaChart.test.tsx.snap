// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PizzaChart > renders PizzaChart component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="pizza-chart"
  id="pizza-chart"
>
  <g
    transform="translate(0,0)"
  >
    <g>
      <path
        class="fill-gray-200"
        d="M0,0Z"
      />
      <path
        class="fill-gray-200"
        d="M0,0Z"
      />
      <path
        class="fill-gray-200"
        d="M0,0Z"
      />
      <path
        class="fill-gray-200"
        d="M0,0Z"
      />
      <path
        class="fill-gray-200"
        d="M0,0Z"
      />
      <path
        class="fill-gray-200"
        d="M0,0Z"
      />
      <path
        class="fill-gray-200"
        d="M0,0Z"
      />
    </g>
    <g>
      <path
        class="fill-gray-800"
        d="M0,0Z"
      />
      <path
        class="fill-gray-800"
        d="M0,0Z"
      />
      <path
        class="fill-gray-800"
        d="M0,0Z"
      />
      <path
        class="fill-gray-800"
        d="M0,0Z"
      />
      <path
        class="fill-gray-800"
        d="M0,0Z"
      />
      <path
        class="fill-gray-800"
        d="M0,0Z"
      />
      <path
        class="fill-gray-800"
        d="M0,0Z"
      />
    </g>
  </g>
</svg>
`;
